minigunapp/tasks.py

```diff
@@ -0,0 +1,40 @@
+import logging
+
+from celery import shared_task
+from minigunapp.models import Email
+from django.core.mail import send_mail
+
+
+@shared_task
+def send_email(id):
+    """ Task that attempts to send the email with the given id.
+    Retries on a delay until max failures reached.
+    """
+
+    email = Email.objects.get(id=id)
+
+    try:
+        send_mail(
+            email.subject,
+            email.body,
+            email.from_,
+            [email.to],
+            fail_silently=False,
+        )
+        email.status = 'sent'
+        logging.info("Sent email {0}".format(email.id))
```

Comment:
If the celery worker gets killed anywhere in the task, it will never log the successful completion of the task. I suspect it would attempt the task again, possibly resulting in a duplicate email, but I think the details would depend on how celery marks its tasks as reserved or not. If the dead worker left the task reserved, maybe nobody would pick it up again. This is an atomicity issue: I want to do A(send) and B(record that I have sent) and there's a critical section encompassing A and B together and I'll have a bad time if I do one but not the other.

So some options:

* Mark the start of your task with an atomic operation in your db. Do the same when its completed. Write code to handle the case of something starting but never completing, maybe by attempting to resume where it left off
* Reduce the size of the critical section
* Adopt a fancy two-phase commit strategy